Memory leak when running queue after about 15 jobs

### Problem description:
1. Start up Handbrake
2. Load job using presets to produce mkv file with AC3 audio using H.264 (intel QSV)
3. Run jobs individually - no problems.
4. Add jobs to queue (about 30) and start the queue.
5. After about 15 or so jobs being run successfully in the queue, handbrake crashes with out of memory.

**I suspect it could have something to do with H.264 Intel QSV encoder calls which uses the D3D11 implementation, as in the past I had used the slower H.264 software encoder and had not see such problems with Handbrake.  Though that could be because it took so much longer to do the encoding I only queued fewer jobs.**

**_UPDATE: So not only is there a memory leak, but when it approached the magic 4GB (32bit counter limit), bang it crashes._**
 
### HandBrake version (e.g., 1.3.0):
1.3.3

### Operating system and version:
Win10 x64 Pro 1903

### Error message text and/or screenshot:
Handbrake has crashed (program complete closed), the only info is from the log for the last job processed
[03:41:07] qsv_enc_init: using 'hardware (1) via D3D11' implementation, API: 1.30
out of memory

### HandBrake Activity Log:
~~~
HandBrake 1.3.3 (2020061300)
OS: Microsoft Windows NT 10.0.18362.0
CPU: Intel(R) Core(TM) i5-6600 CPU @ 3.30GHz
Ram: 16296 MB, 
GPU Information:
  Intel(R) HD Graphics 530 - 26.20.100.7263
Screen: 1280x1024
Temp Dir: D:\AppData\Local\Temp\
Install Dir: C:\Program Files\HandBrake
Data Dir: D:\AppData\Roaming\HandBrake

-------------------------------------------


# Starting Encode ...

[03:41:05] hb_init: starting libhb thread
[03:41:05] Starting work at: Fri Jun 26 03:41:05 2020
[03:41:05] 1 job(s) to process
[03:41:05] json job:
{
  "Audio": {
    "AudioList": [
      {
        "Bitrate": 384,
        "DRC": 0.0,
        "Encoder": "ac3",
        "Gain": 0.0,
        "Mixdown": 7,
        "NormalizeMixLevel": false,
        "Samplerate": 48000,
        "Track": 0,
        "DitherMethod": 0
      }
    ],
    "CopyMask": [
      "copy:ac3"
    ],
    "FallbackEncoder": "ac3"
  },
  "Destination": {
    "ChapterList": [],
    "ChapterMarkers": false,
    "AlignAVStart": false,
    "File": "E:\\TV\\WhenGoesAround CRF22.mkv",
    "Mp4Options": {
      "IpodAtom": false,
      "Mp4Optimize": false
    },
    "Mux": "av_mkv"
  },
  "Filters": {
    "FilterList": [
      {
        "ID": 4,
        "Settings": {
          "mode": "7"
        }
      },
      {
        "ID": 3,
        "Settings": {
          "block-height": "16",
          "block-thresh": "40",
          "block-width": "16",
          "filter-mode": "2",
          "mode": "3",
          "motion-thresh": "1",
          "spatial-metric": "2",
          "spatial-thresh": "1"
        }
      },
      {
        "ID": 12,
        "Settings": {
          "crop-bottom": "0",
          "crop-left": "0",
          "crop-right": "0",
          "crop-top": "0",
          "height": "1080",
          "width": "1920"
        }
      },
      {
        "ID": 6,
        "Settings": {
          "mode": "2",
          "rate": "27000000/1126125"
        }
      }
    ]
  },
  "PAR": {
    "Num": 1,
    "Den": 1
  },
  "Metadata": {},
  "SequenceID": 0,
  "Source": {
    "Angle": 1,
    "Range": {
      "Type": "chapter",
      "Start": 1,
      "End": 1
    },
    "Title": 1,
    "Path": "D:\\WhenGoesAround.mkv"
  },
  "Subtitle": {
    "Search": {
      "Burn": false,
      "Default": false,
      "Enable": false,
      "Forced": false
    },
    "SubtitleList": []
  },
  "Video": {
    "Encoder": "qsv_h264",
    "Level": "4.1",
    "TwoPass": false,
    "Turbo": false,
    "ColorMatrixCode": 0,
    "Options": "",
    "Preset": "quality",
    "Profile": "main",
    "Quality": 22.0,
    "QSV": {
      "Decode": false,
      "AsyncDepth": 0
    }
  }
}
[03:41:05] CPU: Intel(R) Core(TM) i5-6600 CPU @ 3.30GHz
[03:41:05]  - Intel microarchitecture Skylake
[03:41:05]  - logical processor count: 4
[03:41:05] Intel Quick Sync Video support: yes
[03:41:05]  - Intel Media SDK hardware: API 1.30 (minimum: 1.3)
[03:41:05]  - H.264 encoder: yes
[03:41:05]     - preferred implementation: hardware (any) via D3D11
[03:41:05]     - capabilities (hardware):  breftype icq+la+i+downs vsinfo opt1 opt2+mbbrc+extbrc+trellis+ib_adapt+nmpslice
[03:41:05]  - H.265 encoder: yes (8bit: yes, 10bit: no)
[03:41:05]     - preferred implementation: hardware (any) via D3D11
[03:41:05]     - capabilities (hardware):  bpyramid icq vsinfo opt1
[03:41:05] hb_scan: path=D:\ WhenGoesAround.mkv, title_index=1
udfread ERROR: ECMA 167 Volume Recognition failed
src/libbluray/disc/disc.c:323: failed opening UDF image D:\WhenGoesAround.mkv
src/libbluray/disc/disc.c:424: error opening file BDMV\index.bdmv
src/libbluray/disc/disc.c:424: error opening file BDMV\BACKUP\index.bdmv
src/libbluray/bluray.c:2585: nav_get_title_list(D:\WhenGoesAround.mkv\) failed
[03:41:05] bd: not a bd - trying as a stream/file instead
libdvdnav: Using dvdnav version 6.0.1
libdvdread: Encrypted DVD support unavailable.
libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.IFO failed
libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.BUP failed
libdvdread: Can't open file VIDEO_TS.IFO.
libdvdnav: vm: failed to read VIDEO_TS.IFO
[03:41:05] dvd: not a dvd - trying as a stream/file instead
Input #0, matroska,webm, from 'D:\WhenGoesAround.mkv':
  Metadata:
    encoder         : libebml v1.3.9 + libmatroska v1.5.2
    creation_time   : 2020-06-22T11:07:23.000000Z
  Duration: 00:42:53.82, start: 0.000000, bitrate: 8831 kb/s
    Stream #0:0(eng): Video: h264 (High), yuv420p(progressive), 1920x1080 [SAR 1:1 DAR 16:9], 23.98 fps, 23.98 tbr, 1k tbn, 47.95 tbc (default)
    Metadata:
      BPS-eng         : 8190100
      DURATION-eng    : 00:42:53.782000000
      NUMBER_OF_FRAMES-eng: 61709
      NUMBER_OF_BYTES-eng: 2634941499
      _STATISTICS_WRITING_APP-eng: mkvmerge v35.0.0 ('All The Love In The World') 64-bit
      _STATISTICS_WRITING_DATE_UTC-eng: 2020-06-22 11:07:23
      _STATISTICS_TAGS-eng: BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES
    Stream #0:1(eng): Audio: ac3, 48000 Hz, 5.1(side), fltp, 640 kb/s (default)
    Metadata:
      BPS-eng         : 640000
      DURATION-eng    : 00:42:53.824000000
      NUMBER_OF_FRAMES-eng: 80432
      NUMBER_OF_BYTES-eng: 205905920
      _STATISTICS_WRITING_APP-eng: mkvmerge v35.0.0 ('All The Love In The World') 64-bit
      _STATISTICS_WRITING_DATE_UTC-eng: 2020-06-22 11:07:23
      _STATISTICS_TAGS-eng: BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES
[03:41:05] scan: decoding previews for title 1
[03:41:05] scan: audio 0x1: ac3, rate=48000Hz, bitrate=640000 English (AC3) (5.1 ch) (640 kbps)
[03:41:06] scan: 10 previews, 1920x1080, 23.976 fps, autocrop = 0/0/0/0, aspect 16:9, PAR 1:1
[03:41:06] scan: supported video decoders: avcodec qsv
[03:41:06] libhb: scan thread found 1 valid title(s)
[03:41:06] Starting Task: Encoding Pass
[03:41:06] Skipping crop/scale filter
[03:41:06] job configuration:
[03:41:06]  * source
[03:41:06]    + D:\WhenGoesAround.mkv
[03:41:06]    + title 1, chapter(s) 1 to 1
[03:41:06]    + container: matroska,webm
[03:41:06]    + data rate: 8831 kbps
[03:41:06]  * destination
[03:41:06]    + E:\WhenGoesAround CRF22.mkv
[03:41:06]    + container: Matroska (libavformat)
[03:41:06]  * video track
[03:41:06]    + decoder: h264
[03:41:06]    + filters
[03:41:06]      + Comb Detect (mode=3:spatial-metric=2:motion-thresh=1:spatial-thresh=1:filter-mode=2:block-thresh=40:block-width=16:block-height=16)
[03:41:06]      + Decomb (mode=39)
[03:41:06]      + Framerate Shaper (mode=2:rate=27000000/1126125)
[03:41:06]        + frame rate: 23.976 fps -> peak rate limited to 23.976 fps
[03:41:06]    + Output geometry
[03:41:06]      + storage dimensions: 1920 x 1080
[03:41:06]      + pixel aspect ratio: 1 : 1
[03:41:06]      + display dimensions: 1920 x 1080
[03:41:06]    + encoder: H.264 (Intel Media SDK)
[03:41:06]      + preset:  quality
[03:41:06]      + profile: main
[03:41:06]      + level:   4.1
[03:41:06]      + quality: 22.00 (ICQ)
[03:41:06]      + color profile: 1-1-1
[03:41:06]  * audio track 1
[03:41:06]    + decoder: English (AC3) (5.1 ch) (640 kbps) (track 1, id 0x1)
[03:41:06]      + bitrate: 640 kbps, samplerate: 48000 Hz
[03:41:06]    + mixdown: 5.1 Channels
[03:41:06]    + dither: none
[03:41:06]    + encoder: AC3 (libavcodec)
[03:41:06]      + bitrate: 384 kbps, samplerate: 48000 Hz
[03:41:06] sync: expecting 61710 video frames
[03:41:06] encqsvInit: using encode-only path
[03:41:06] encqsvInit: H.264/AVC Main profile @ level 4.1
[03:41:06] encqsvInit: TargetUsage 2 AsyncDepth 3
[03:41:06] encqsvInit: GopRefDist 3 GopPicSize 24 NumRefFrame 3
[03:41:06] encqsvInit: BFramesMax 2 BRefType off
[03:41:06] encqsvInit: RateControlMethod LA_ICQ ICQQuality 22 LookAheadDepth 40
[03:41:06] sync: first pts video is 0
[03:41:06] sync: Chapter 1 at frame 1 time 0
[03:41:06] sync: first pts audio 0x1 is 0
**[03:41:07] qsv_enc_init: using 'hardware (1) via D3D11' implementation, API: 1.30
out of memory**

~~~

--- a/libhb/enc_qsv.c
+++ b/libhb/enc_qsv.c
@@ -1765,7 +1765,7 @@ void encqsvClose(hb_work_object_t *w)
             }
 
             /* QSV context cleanup and MFXClose */
-            hb_qsv_context_clean(qsv_ctx);
+            hb_qsv_context_clean(qsv_ctx,hb_qsv_full_path_is_enabled(pv->job));
 
             hb_display_close(&pv->display);
 

--- a/libhb/handbrake/qsv_libav.h
+++ b/libhb/handbrake/qsv_libav.h
@@ -460,7 +460,7 @@ void hb_qsv_dts_pop(hb_qsv_context *);
 
 hb_qsv_stage *hb_qsv_stage_init(void);
 void hb_qsv_stage_clean(hb_qsv_stage **);
-int hb_qsv_context_clean(hb_qsv_context *);
+int hb_qsv_context_clean(hb_qsv_context *, int);
 
 int ff_qsv_is_sync_in_pipe(mfxSyncPoint *, hb_qsv_context *);
 int ff_qsv_is_surface_in_pipe(mfxFrameSurface1 *, hb_qsv_context *);

--- a/libhb/qsv_filter.c
+++ b/libhb/qsv_filter.c
@@ -462,7 +462,7 @@ static void hb_qsv_filter_close( hb_filter_object_t * filter )
         qsv_filter_close(qsv,HB_QSV_VPP_DEFAULT);
 
         // closing the commong stuff
-        hb_qsv_context_clean(qsv);
+        hb_qsv_context_clean(qsv,hb_qsv_full_path_is_enabled(pv->job));
     }
     hb_buffer_list_close(&pv->list);
     free( pv );

--- a/libhb/qsv_filter_pp.c
+++ b/libhb/qsv_filter_pp.c
@@ -36,7 +36,7 @@ THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "handbrake/qsv_filter.h"
 #include "handbrake/qsv_libav.h"
 #include "handbrake/qsv_memory.h"
-
+#include "handbrake/qsv_common.h"
 
 static int hb_qsv_filter_pre_init( hb_filter_object_t * filter,
                                hb_filter_init_t * init );
@@ -520,7 +520,7 @@ static void hb_qsv_filter_pre_close( hb_filter_object_t * filter ){
         qsv_filter_close(qsv,HB_QSV_VPP_USER);
 
         // closing the commong stuff
-        hb_qsv_context_clean(qsv);
+        hb_qsv_context_clean(qsv,hb_qsv_full_path_is_enabled(pv->job));
     }
     hb_cond_close(&pv->pre.frame_completed);
     hb_lock_close(&pv->pre.frame_completed_lock);

--- a/libhb/qsv_libav.c
+++ b/libhb/qsv_libav.c
@@ -207,7 +207,7 @@ void hb_qsv_add_context_usage(hb_qsv_context * qsv, int is_threaded)
     }
 }
 
-int hb_qsv_context_clean(hb_qsv_context * qsv)
+int hb_qsv_context_clean(hb_qsv_context * qsv, int full_job)
 {
     int is_active = 0;
     mfxStatus sts = MFX_ERR_NONE;
@@ -235,10 +235,9 @@ int hb_qsv_context_clean(hb_qsv_context * qsv)
         if (qsv->pipes)
             hb_qsv_pipe_list_clean(&qsv->pipes);
 
-        if (qsv->mfx_session) {
-            // Intentionally leave it because the decode session closed first and encoder session hangs
-            //sts = MFXClose(qsv->mfx_session);
-            //HB_QSV_CHECK_RESULT(sts, MFX_ERR_NONE, sts);
+        if (qsv->mfx_session && !full_job) {
+            sts = MFXClose(qsv->mfx_session);
+            HB_QSV_CHECK_RESULT(sts, MFX_ERR_NONE, sts);
             qsv->mfx_session = 0;
         }
     }

